include/npcomp/RefBackend/Runtime/UserAPI.h

```diff
@@ -150,6 +187,131 @@ class Tensor {
   // Sizes are tail-allocated.
 };
 
+// RtValue is a generic tagged union used to hold all value types
+// The tag determines the type, and the payload represents the stored
+// contents of an object. If an object is not trivially destructible,
+// then it must be refcounted and must have a refCount.
+#define NPCOMP_FORALL_TAGS(_)                                                  \
+  _(None)                                                                      \
+  _(Bool)                                                                      \
+  _(Int)                                                                       \
+  _(Double)                                                                    \
+  _(Tensor)
+
+struct RtValue final {
+
+  RtValue() : payload{0}, tag(Tag::None) {}
+
+  // Bool
+  RtValue(bool b) : tag(Tag::Bool) { payload.asBool = b; }
+  bool isBool() const { return Tag::Bool == tag; }
+  bool toBool() const {
+    assert(isBool());
+    return payload.asBool;
+  }
+
+  // Int
+  RtValue(std::int64_t i) : tag(Tag::Int) { payload.asInt = i; }
+  RtValue(std::int32_t i) : RtValue(static_cast<int64_t>(i)) {}
+  bool isInt() const { return Tag::Int == tag; }
+  bool toInt() const {
+    assert(isInt());
+    return payload.asInt;
+  }
+
+  // Double
+  RtValue(double d) : tag(Tag::Double) { payload.asDouble = d; }
+  bool isDouble() const { return Tag::Double == tag; }
+  bool toDouble() const {
+    assert(isDouble());
+    return payload.asDouble;
+  }
+
+  // Tensor
+  RtValue(Ref<Tensor> tensor) : tag(Tag::Tensor) {
+    payload.asRefTargetPtr = reinterpret_cast<RefTarget *>(tensor.takePtr());
+  }
+  bool isTensor() const { return Tag::Tensor == tag; }
+  Ref<Tensor> toTensor() const {
+    assert(isTensor());
+    return Ref<Tensor>(reinterpret_cast<Tensor *>(payload.asRefTargetPtr));
+  }
+
+  // Ref
+  bool isRef() const { return isTensor(); }
+
+  // RtValue (downcast)
+  const RtValue &toRtValue() const { return *this; }
+  RtValue &toRtValue() { return *this; }
+
+  // Stringify tag for debugging.
+  std::string tagKind() const {
+    switch (tag) {
+#define DEFINE_CASE(x)                                                         \
+  case Tag::x:                                                                 \
+    return #x;
+      NPCOMP_FORALL_TAGS(DEFINE_CASE)
+#undef DEFINE_CASE
+    }
+    // TODO(brycearden): Print tag here
+    return "InvalidTag!";
+  }
+
+  RtValue(const RtValue &rhs) : RtValue(rhs.payload, rhs.tag) {
+    if (isRef()) {
+      payload.asRefTargetPtr->incref();
+    }
+  }
+  RtValue(RtValue &&rhs) noexcept : RtValue() { swap(rhs); }
+
+  RtValue &operator=(RtValue &&rhs) & noexcept {
+    RtValue(std::move(rhs)).swap(*this); // this also sets rhs to None
+    return *this;
+  }
+  RtValue &operator=(RtValue const &rhs) & {
+    RtValue(rhs).swap(*this);
+    return *this;
+  }
+
+  ~RtValue() {
+    if (isRef()) {
+      if (isTensor()) {
+        auto raii = Ref<Tensor>::reclaimPtr(
```

Comment:
I don't think you need reclaimPtr. You just need `Ref<Tensor>::decref(static_cast<Tensor *>(payload.asRefTargetPtr))`.

(you shouldn't need RefTarget with this approach -- the point of RefTarget is for it to have a virtual method releaseResources which allows doing destructor chaining without checking isTensor or casting concretely to Tensor).